---

_You are currently looking at **version 1.2** of this notebook. To download notebooks and datafiles, as well as get help on Jupyter notebooks in the Coursera platform, visit the [Jupyter Notebook FAQ](https://www.coursera.org/learn/python-social-network-analysis/resources/yPcBs) course resource._

---

# Assignment 4

In [1]:
import networkx as nx
import pandas as pd
import numpy as np
import pickle

---

## Part 1 - Random Graph Identification

For the first part of this assignment you will analyze randomly generated graphs and determine which algorithm created them.

In [2]:
P1_Graphs = pickle.load(open('A4_graphs','rb'))
P1_Graphs

<br>
`P1_Graphs` is a list containing 5 networkx graphs. Each of these graphs were generated by one of three possible algorithms:
* Preferential Attachment (`'PA'`)
* Small World with low probability of rewiring (`'SW_L'`)
* Small World with high probability of rewiring (`'SW_H'`)

Anaylze each of the 5 graphs and determine which of the three algorithms generated the graph.

*The `graph_identification` function should return a list of length 5 where each element in the list is either `'PA'`, `'SW_L'`, or `'SW_H'`.*

In [3]:
def degree_distribution(G):
    degrees = G.degree()
    degree_values = sorted(set(degrees.values()))
    histogram = [list(degrees.values()).count(i)/float(nx.number_of_nodes( G)) for i in degree_values]
    return histogram

In [4]:
for G in P1_Graphs:
    print((max(degree_distribution(G)), min(degree_distribution(G)), len(degree_distribution(G)), nx.average_clustering(G), nx.average_shortest_path_length(G)))

(0.465, 0.001, 29, 0.03167539146454044, 4.099161161161161)
(0.615, 0.001, 7, 0.5642419635919628, 5.089871871871872)
(0.756, 0.0026666666666666666, 5, 0.4018222222222227, 9.378702269692925)
(0.3453333333333333, 0.0013333333333333333, 37, 0.03780379975223251, 3.1048046283934134)
(0.2906666666666667, 0.0013333333333333333, 9, 0.0033037037037037037, 5.0785509568313305)


In [5]:
def graph_identification():
    
    # Your Code Here
    
    return ['PA', 'SW_L', 'SW_L', 'PA', 'SW_H']

---

## Part 2 - Company Emails

For the second part of this assignment you will be workking with a company's email network where each node corresponds to a person at the company, and each edge indicates that at least one email has been sent between two people.

The network also contains the node attributes `Department` and `ManagementSalary`.

`Department` indicates the department in the company which the person belongs to, and `ManagementSalary` indicates whether that person is receiving a management position salary.

In [6]:
G = nx.read_gpickle('email_prediction.txt')

print(nx.info(G))

Name: 
Type: Graph
Number of nodes: 1005
Number of edges: 16706
Average degree:  33.2458


In [7]:
G.nodes(data=True)

[(0, {'Department': 1, 'ManagementSalary': 0.0}),
 (1, {'Department': 1, 'ManagementSalary': nan}),
 (2, {'Department': 21, 'ManagementSalary': nan}),
 (3, {'Department': 21, 'ManagementSalary': 1.0}),
 (4, {'Department': 21, 'ManagementSalary': 1.0}),
 (5, {'Department': 25, 'ManagementSalary': nan}),
 (6, {'Department': 25, 'ManagementSalary': 1.0}),
 (7, {'Department': 14, 'ManagementSalary': 0.0}),
 (8, {'Department': 14, 'ManagementSalary': nan}),
 (9, {'Department': 14, 'ManagementSalary': 0.0}),
 (10, {'Department': 9, 'ManagementSalary': 0.0}),
 (11, {'Department': 14, 'ManagementSalary': 0.0}),
 (12, {'Department': 14, 'ManagementSalary': 1.0}),
 (13, {'Department': 26, 'ManagementSalary': 1.0}),
 (14, {'Department': 4, 'ManagementSalary': nan}),
 (15, {'Department': 17, 'ManagementSalary': 0.0}),
 (16, {'Department': 34, 'ManagementSalary': 0.0}),
 (17, {'Department': 1, 'ManagementSalary': 0.0}),
 (18, {'Department': 1, 'ManagementSalary': nan}),
 (19, {'Department': 14, 'Ma

In [8]:
G.edges(data=True)

[(0, 1, {}),
 (0, 17, {}),
 (0, 316, {}),
 (0, 146, {}),
 (0, 581, {}),
 (0, 268, {}),
 (0, 221, {}),
 (0, 218, {}),
 (0, 18, {}),
 (0, 734, {}),
 (0, 178, {}),
 (0, 380, {}),
 (0, 0, {}),
 (0, 459, {}),
 (0, 215, {}),
 (0, 250, {}),
 (0, 148, {}),
 (0, 73, {}),
 (0, 74, {}),
 (0, 248, {}),
 (0, 498, {}),
 (0, 226, {}),
 (0, 101, {}),
 (0, 377, {}),
 (0, 177, {}),
 (0, 103, {}),
 (0, 560, {}),
 (0, 309, {}),
 (0, 88, {}),
 (0, 5, {}),
 (0, 297, {}),
 (0, 313, {}),
 (0, 223, {}),
 (0, 238, {}),
 (0, 368, {}),
 (0, 266, {}),
 (0, 222, {}),
 (0, 283, {}),
 (0, 6, {}),
 (0, 64, {}),
 (0, 65, {}),
 (0, 166, {}),
 (0, 120, {}),
 (1, 74, {}),
 (1, 17, {}),
 (1, 316, {}),
 (1, 1, {}),
 (1, 268, {}),
 (1, 495, {}),
 (1, 377, {}),
 (1, 549, {}),
 (1, 310, {}),
 (1, 218, {}),
 (1, 215, {}),
 (1, 147, {}),
 (1, 106, {}),
 (1, 21, {}),
 (1, 225, {}),
 (1, 82, {}),
 (1, 254, {}),
 (1, 155, {}),
 (1, 146, {}),
 (1, 85, {}),
 (1, 284, {}),
 (1, 189, {}),
 (1, 250, {}),
 (1, 726, {}),
 (1, 548, {}),
 (

In [9]:
# Initialize the dataframe, using the nodes as the index
df = pd.DataFrame(index=G.nodes())

In [10]:
df['Department']=pd.Series(nx.get_node_attributes(G, 'Department'))
df['clustering'] = pd.Series(nx.clustering(G))
df['degree'] = pd.Series(G.degree())
df['ManagementSalary']=pd.Series(nx.get_node_attributes(G, 'ManagementSalary'))

In [11]:
df.head()

Department  clustering  degree  ManagementSalary
0           1    0.276423      44               0.0
1           1    0.265306      52               NaN
2          21    0.297803      95               NaN
3          21    0.384910      71               1.0
4          21    0.318691      96               1.0

In [12]:
test_dataset = df.loc[pd.isnull(df['ManagementSalary'])]

In [13]:
train_dataset = df.loc[~pd.isnull(df['ManagementSalary'])]

In [14]:
train_dataset.head()

Department  clustering  degree  ManagementSalary
0           1    0.276423      44               0.0
3          21    0.384910      71               1.0
4          21    0.318691      96               1.0
6          25    0.155183     115               1.0
7          14    0.287785      72               0.0

In [15]:
from sklearn.ensemble import RandomForestClassifier

In [16]:
X_train = train_dataset.drop('ManagementSalary', axis=1)
y_train = train_dataset.ManagementSalary

clf = RandomForestClassifier().fit(X_train, y_train)

In [17]:

from sklearn.metrics import confusion_matrix

confusion_matrix(y_train, clf.predict(X_train))

array([[633,   1],
       [  8, 111]])

In [18]:
probs = clf.predict_proba(test_dataset[list(X_train.columns)])

In [19]:
probs = probs[:,1]

In [20]:
probs

array([ 0. ,  0.9,  1. ,  0. ,  0.1,  0.1,  0. ,  0.7,  0.5,  0.1,  0. ,
        0. ,  0. ,  0.7,  0.5,  0.9,  1. ,  0.9,  0. ,  0. ,  0. ,  0. ,
        0.9,  0.1,  0.5,  0. ,  0.1,  1. ,  0. ,  0.7,  0.7,  0.7,  0. ,
        0. ,  0.1,  0.2,  0.2,  0.4,  0. ,  0.3,  0. ,  0.2,  0. ,  0.4,
        0.1,  0. ,  0. ,  0. ,  1. ,  0.7,  0.9,  0.6,  0.7,  1. ,  0. ,
        0.1,  0.3,  0. ,  0. ,  0.2,  0. ,  0. ,  0. ,  1. ,  0.1,  0.8,
        0. ,  0. ,  0.1,  0. ,  0.1,  0. ,  0. ,  0. ,  0. ,  0. ,  0. ,
        0.1,  0. ,  0.1,  0. ,  0. ,  0.3,  0.2,  0.8,  0.9,  0.7,  0.6,
        0.3,  0. ,  0. ,  0.2,  0.1,  0.2,  0.3,  0. ,  0. ,  0. ,  0.4,
        0.7,  0.2,  1. ,  1. ,  0. ,  0.6,  0. ,  0. ,  0.2,  0. ,  0. ,
        0. ,  0.2,  0.1,  0. ,  0. ,  0. ,  0. ,  0.1,  0. ,  0.5,  0.9,
        0. ,  0.4,  0.1,  0. ,  0.1,  0. ,  0.1,  0.5,  0.1,  0. ,  0.1,
        0. ,  0.5,  1. ,  0. ,  0. ,  0. ,  0. ,  0. ,  0. ,  0. ,  0. ,
        0.1,  0. ,  0. ,  0.3,  0. ,  0.2,  0. ,  0

In [21]:
series_probs =pd.Series(probs, index=test_dataset.index)

In [22]:
series_probs

1       0.0
2       0.9
5       1.0
8       0.0
14      0.1
18      0.1
27      0.0
30      0.7
31      0.5
34      0.1
37      0.0
40      0.0
45      0.0
54      0.7
55      0.5
60      0.9
62      1.0
65      0.9
77      0.0
79      0.0
97      0.0
101     0.0
103     0.9
108     0.1
113     0.5
122     0.0
141     0.1
142     1.0
144     0.0
145     0.7
       ... 
913     0.1
914     0.0
915     0.0
918     0.3
923     0.0
926     0.0
931     0.0
934     0.0
939     0.0
944     0.0
945     0.0
947     0.0
950     0.0
951     0.0
953     0.0
959     0.0
962     0.0
963     0.0
968     0.0
969     0.0
974     0.0
984     0.0
987     0.0
989     0.0
991     0.1
992     0.0
994     0.0
996     0.0
1000    0.1
1001    0.0
Length: 252, dtype: float64

### Part 2A - Salary Prediction

Using network `G`, identify the people in the network with missing values for the node attribute `ManagementSalary` and predict whether or not these individuals are receiving a management position salary.

To accomplish this, you will need to create a matrix of node features using networkx, train a sklearn classifier on nodes that have `ManagementSalary` data, and predict a probability of the node receiving a management salary for nodes where `ManagementSalary` is missing.



Your predictions will need to be given as the probability that the corresponding employee is receiving a management position salary.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 252 with the data being the probability of receiving management salary, and the index being the node id.

    Example:
    
        1       1.0
        2       0.0
        5       0.8
        8       1.0
            ...
        996     0.7
        1000    0.5
        1001    0.0
        Length: 252, dtype: float64

In [23]:
def salary_predictions():
    
    # Your Code Here
    
    return series_probs

In [24]:
salary_predictions()

1       0.0
2       0.9
5       1.0
8       0.0
14      0.1
18      0.1
27      0.0
30      0.7
31      0.5
34      0.1
37      0.0
40      0.0
45      0.0
54      0.7
55      0.5
60      0.9
62      1.0
65      0.9
77      0.0
79      0.0
97      0.0
101     0.0
103     0.9
108     0.1
113     0.5
122     0.0
141     0.1
142     1.0
144     0.0
145     0.7
       ... 
913     0.1
914     0.0
915     0.0
918     0.3
923     0.0
926     0.0
931     0.0
934     0.0
939     0.0
944     0.0
945     0.0
947     0.0
950     0.0
951     0.0
953     0.0
959     0.0
962     0.0
963     0.0
968     0.0
969     0.0
974     0.0
984     0.0
987     0.0
989     0.0
991     0.1
992     0.0
994     0.0
996     0.0
1000    0.1
1001    0.0
Length: 252, dtype: float64

### Part 2B - New Connections Prediction

For the last part of this assignment, you will predict future connections between employees of the network. The future connections information has been loaded into the variable `future_connections`. The index is a tuple indicating a pair of nodes that currently do not have a connection, and the `Future Connection` column indicates if an edge between those two nodes will exist in the future, where a value of 1.0 indicates a future connection.

In [25]:
future_connections = pd.read_csv('Future_Connections.csv', index_col=0, converters={0: eval})
future_connections.head(10)

Future Connection
(6, 840)                  0.0
(4, 197)                  0.0
(620, 979)                0.0
(519, 872)                0.0
(382, 423)                0.0
(97, 226)                 1.0
(349, 905)                0.0
(429, 860)                0.0
(309, 989)                0.0
(468, 880)                0.0

In [26]:
# Initialize the dataframe, using the edges as the index
#df = pd.DataFrame(index=G.edges())
preferential_attachment = list(nx.preferential_attachment(G))
df = pd.DataFrame(index=[(x[0], x[1]) for x in preferential_attachment])

In [27]:
df['preferential attachment'] = [i[2] for i in nx.preferential_attachment(G, df.index)]
df['Common Neighbors'] = df.index.map(lambda emp: len(list(nx.common_neighbors(G, emp[0], emp[1]))))
df['jaccard_coefficient'] = [i[2] for i in nx.jaccard_coefficient(G,  df.index)]
df['resource_allocation_index'] = [x[2] for x in nx.resource_allocation_index(G, df.index)]

In [28]:
df.shape

(488446, 4)

In [29]:
future_connections.shape

(488446, 1)

In [30]:
full_df = future_connections.join(df,how='left')

In [31]:
full_df.shape

(488446, 5)

In [32]:
full_df.head()

Future Connection  preferential attachment  Common Neighbors  \
(6, 840)                  0.0                     2070                 9   
(4, 197)                  0.0                     3552                 2   
(620, 979)                0.0                       28                 0   
(519, 872)                0.0                      299                 2   
(382, 423)                0.0                      205                 0   

            jaccard_coefficient  resource_allocation_index  
(6, 840)               0.073770                   0.136721  
(4, 197)               0.015504                   0.008437  
(620, 979)             0.000000                   0.000000  
(519, 872)             0.060606                   0.039726  
(382, 423)             0.000000                   0.000000

In [33]:
test_dataset = full_df.loc[pd.isnull(full_df['Future Connection'])]
train_dataset = full_df.loc[~pd.isnull(full_df['Future Connection'])]

In [34]:
test_dataset.fillna(0, inplace=True)
train_dataset.fillna(0, inplace=True)

/opt/conda/lib/python3.6/site-packages/pandas/core/frame.py:2754: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  downcast=downcast, **kwargs)


In [35]:
X_train = train_dataset.drop('Future Connection', axis=1)
y_train = train_dataset['Future Connection']



clf = RandomForestClassifier().fit(X_train, y_train)

In [36]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_train, clf.predict(X_train))

array([[336791,    211],
       [  4622,  24710]])

In [37]:
probs = clf.predict_proba(test_dataset[list(X_train.columns)])
probs = probs[:,1]
series_probs =pd.Series(probs, index=test_dataset.index)
series_probs

(107, 348)    0.000000
(542, 751)    0.013090
(20, 426)     0.800000
(50, 989)     0.012202
(942, 986)    0.011552
(324, 857)    0.016535
(13, 710)     0.200000
(19, 271)     0.200000
(319, 878)    0.013260
(659, 707)    0.014111
(49, 843)     0.014260
(208, 893)    0.006808
(377, 469)    0.000000
(405, 999)    0.000000
(129, 740)    0.000000
(292, 618)    0.000000
(239, 689)    0.005965
(359, 373)    0.150970
(53, 523)     0.000000
(276, 984)    0.006442
(202, 997)    0.009479
(604, 619)    0.100000
(270, 911)    0.013580
(261, 481)    0.000000
(200, 450)    0.400000
(213, 634)    0.016706
(644, 735)    0.200000
(346, 553)    0.000000
(521, 738)    0.019821
(422, 953)    0.000000
                ...   
(672, 848)    0.013580
(28, 127)     1.000000
(202, 661)    0.014572
(54, 195)     1.000000
(295, 864)    0.011981
(814, 936)    0.010584
(839, 874)    0.011552
(139, 843)    0.013503
(461, 544)    0.009771
(68, 487)     0.000000
(622, 932)    0.017946
(504, 936)    0.000000
(479, 528) 

Using network `G` and `future_connections`, identify the edges in `future_connections` with missing values and predict whether or not these edges will have a future connection.

To accomplish this, you will need to create a matrix of features for the edges found in `future_connections` using networkx, train a sklearn classifier on those edges in `future_connections` that have `Future Connection` data, and predict a probability of the edge being a future connection for those edges in `future_connections` where `Future Connection` is missing.



Your predictions will need to be given as the probability of the corresponding edge being a future connection.

The evaluation metric for this assignment is the Area Under the ROC Curve (AUC).

Your grade will be based on the AUC score computed for your classifier. A model which with an AUC of 0.88 or higher will receive full points, and with an AUC of 0.82 or higher will pass (get 80% of the full points).

Using your trained classifier, return a series of length 122112 with the data being the probability of the edge being a future connection, and the index being the edge as represented by a tuple of nodes.

    Example:
    
        (107, 348)    0.35
        (542, 751)    0.40
        (20, 426)     0.55
        (50, 989)     0.35
                  ...
        (939, 940)    0.15
        (555, 905)    0.35
        (75, 101)     0.65
        Length: 122112, dtype: float64

In [38]:
def new_connections_predictions():
    
    # Your Code Here
    
    return series_probs